src/core/lv_obj_draw_cache.c

```diff
@@ -0,0 +1,226 @@
+/**
+ * @file lv_obj_draw_cache.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "lv_obj_draw_cache.h"
+#include "../others/snapshot/lv_snapshot.h"
+
+/*********************
+ *      DEFINES
+ *********************/
+#if LV_USE_OBJ_DRAW_CACHE
+
+#if !LV_USE_SNAPSHOT
+    # error "Need to enable 'LV_USE_SNAPSHOT'"
+#endif
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+typedef struct _lv_obj_draw_cache_t {
+    lv_img_dsc_t img_dsc;
+    lv_img_cf_t cf;
+    bool invalid;
+    bool skip_cache;
+} lv_obj_draw_cache_t;
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+
+/**********************
+ *  STATIC VARIABLES
+ **********************/
+
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+void lv_obj_draw_cache_set_enable(lv_obj_t * obj, bool en)
+{
+    LV_ASSERT_NULL(obj);
+
+    if(en) {
+        lv_obj_allocate_spec_attr(obj);
+
+        if(obj->spec_attr->draw_cache == NULL) {
+            lv_obj_draw_cache_t * draw_cache = lv_malloc(sizeof(lv_obj_draw_cache_t));
+            LV_ASSERT_MALLOC(draw_cache);
+
+            if(draw_cache == NULL) {
+                return;
+            }
+
+            lv_memset(draw_cache, 0, sizeof(lv_obj_draw_cache_t));
+            draw_cache->cf = LV_IMG_CF_TRUE_COLOR_ALPHA;
+            obj->spec_attr->draw_cache = draw_cache;
+        }
+    }
+    else {
+        _lv_obj_draw_cache_free(obj);
+    }
+
+    lv_obj_invalidate(obj);
+}
+
+void lv_obj_draw_cache_set_img_cf(lv_obj_t * obj, lv_img_cf_t cf)
+{
+    LV_ASSERT_NULL(obj);
+    if(!lv_obj_has_draw_cache(obj)) {
+        return;
+    }
+
+    obj->spec_attr->draw_cache->cf = cf;
+}
+
+lv_img_cf_t lv_obj_draw_cache_get_img_cf(lv_obj_t * obj)
+{
+    LV_ASSERT_NULL(obj);
+    if(!lv_obj_has_draw_cache(obj)) {
+        return LV_IMG_CF_UNKNOWN;
+    }
+
+    return obj->spec_attr->draw_cache->cf;
+}
+
+const lv_img_dsc_t * lv_obj_draw_cache_get_img_dsc(lv_obj_t * obj)
+{
+    LV_ASSERT_NULL(obj);
+    if(!lv_obj_has_draw_cache(obj)) {
+        return NULL;
+    }
+
+    return &obj->spec_attr->draw_cache->img_dsc;
+}
+
+void lv_obj_draw_cache_invalidate(lv_obj_t * obj)
+{
+    LV_ASSERT_NULL(obj);
+    if(!lv_obj_has_draw_cache(obj)) {
+        return;
+    }
+
+    obj->spec_attr->draw_cache->invalid = true;
+}
+
+void lv_obj_draw_cache_invalidate_cancel(lv_obj_t * obj)
+{
+    LV_ASSERT_NULL(obj);
+    if(!lv_obj_has_draw_cache(obj)) {
+        return;
+    }
+
+    obj->spec_attr->draw_cache->invalid = false;
```

Comment:
I'm not sure this is the right approach. If an object's `draw_cache` was already correctly invalidated (e.g. style changed) and it's immediately followed by a position change the original invalidation will be cancelled too.